components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java

```diff
@@ -0,0 +1,188 @@
+            // Check if the previous grant type is consent refresh token type or not.
+            if (!StringUtils.equals(OAuthConstants.GrantTypes.REFRESH_TOKEN, previousGrantType)) {
+                // If the previous grant type is not a refresh token, then check if it's a consent token or not.
+                if (OIDCClaimUtil.isConsentBasedClaimFilteringApplicable(previousGrantType)) {
+                    accessTokenDO.setIsConsentedToken(true);
+                }
+            } else {
+                /* When previousGrantType == refresh_token, we need to check whether the original grant type
+                 is consented or not. */
+                AccessTokenDO accessTokenDOFromTokenIdentifier = OAuth2Util.getAccessTokenDOFromTokenIdentifier(
+                        validationBean.getAccessToken(), false);
+                accessTokenDO.setIsConsentedToken(accessTokenDOFromTokenIdentifier.isConsentedToken());
+            }
+
+            if (accessTokenDO.isConsentedToken()) {
+                tokReqMsgCtx.setConsentedToken(true);
+            }
+        }
+        return accessTokenDO;
+    }
+
+    private boolean validatePersistedAccessToken(RefreshTokenValidationDataDO validationBean, String clientId)
+            throws IdentityOAuth2Exception {
+
+        if (validationBean.getAccessToken() == null) {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Invalid Refresh Token provided for Client with Client Id : %s", clientId));
+            }
+            throw new IdentityOAuth2Exception("Persisted access token data not found");
+        }
+        return true;
+    }
+
+    @Override
+    public boolean isLatestRefreshToken(OAuth2AccessTokenReqDTO tokenReq, RefreshTokenValidationDataDO validationBean,
+                                        String userStoreDomain) throws IdentityOAuth2Exception {
+
+        if (log.isDebugEnabled()) {
+            if (IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.REFRESH_TOKEN)) {
+                log.debug(String.format("Evaluating refresh token. Token value(hashed): %s, Token state: %s",
+                        DigestUtils.sha256Hex(tokenReq.getRefreshToken()), validationBean.getRefreshTokenState()));
+            } else {
+                log.debug(String.format("Evaluating refresh token. Token state: %s",
+                        validationBean.getRefreshTokenState()));
+            }
+        }
+        if (!OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE.equals(validationBean.getRefreshTokenState())) {
+            /* if refresh token is not in active state, check whether there is an access token issued with the same
+             * refresh token.
+             */
+            List<AccessTokenDO> accessTokenBeans = getAccessTokenBeans(tokenReq, validationBean, userStoreDomain);
+            for (AccessTokenDO token : accessTokenBeans) {
+                if (tokenReq.getRefreshToken().equals(token.getRefreshToken())
```

Comment:
is this null-safe?

Reply:
This is the refresh token sent in the request for refresh grant itself. Couldnt find a place in the inbound code validating the existence of the parameter hence adding a null check.

Reply:
fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2091/commits/be0223e8df1b84d536c22ea214ef247d84995dcc